In [1]:
import pandas as pd
%matplotlib inline

In [26]:
print("Morning commuter trips - start time between 7:00 and 8:59:")
! bq query --use_legacy_sql=false 'SELECT count(EXTRACT(HOUR FROM start_date)) AS Commuter_Trips FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) > 6 AND EXTRACT(HOUR FROM start_date) < 9'

Morning commuter trips - start time between 7:00 and 8:59:
Waiting on bqjob_r787c93d00095e98e_00000163beffd84c_1 ... (1s) Current status: DONE   
+----------------+
| Commuter_Trips |
+----------------+
|         199995 |
+----------------+


In [27]:
print("Afternoon commuter trips - start time between 16:00 and 18:59:")
! bq query --use_legacy_sql=false 'SELECT count(EXTRACT(HOUR FROM start_date)) AS Commuter_Trips FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) > 15 AND EXTRACT(HOUR FROM start_date) < 19'

Afternoon commuter trips - start time between 16:00 and 18:59:
Waiting on bqjob_r19b37d6f446b2f5e_00000163bf00bf80_1 ... (0s) Current status: DONE   
+----------------+
| Commuter_Trips |
+----------------+
|         299626 |
+----------------+


In [28]:
print("Morning trips < 30 minutes:")
! bq query --use_legacy_sql=false 'SELECT COUNT(duration_sec) AS Morning_Trips FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) < 9 AND EXTRACT(HOUR FROM start_date) > 6 AND duration_sec <= 1800'
print("Afternoon trips < 30 minutes:")
! bq query --use_legacy_sql=false 'SELECT COUNT(duration_sec) AS Afternoon_Trips FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) < 19 AND EXTRACT(HOUR FROM start_date) > 15 AND duration_sec <= 1800'

Morning trips < 30 minutes:
Waiting on bqjob_r224c18e941dbfd52_00000163bf01c52f_1 ... (1s) Current status: DONE   
+---------------+
| Morning_Trips |
+---------------+
|        197646 |
+---------------+
Afternoon trips < 30 minutes:
Waiting on bqjob_r4b4714c7c3e8746b_00000163bf01d8b2_1 ... (1s) Current status: DONE   
+-----------------+
| Afternoon_Trips |
+-----------------+
|          289743 |
+-----------------+


In [53]:
print("Morning commuter trips - start time between 7:00 and 8:59:")
! bq query --use_legacy_sql=false 'SELECT subscriber_type, count(EXTRACT(HOUR FROM start_date)) AS Commuter_Trips FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) > 6 AND EXTRACT(HOUR FROM start_date) < 9 GROUP BY subscriber_type'

print("Afternoon commuter trips - start time between 16:00 and 18:59:")
! bq query --use_legacy_sql=false 'SELECT subscriber_type, count(EXTRACT(HOUR FROM start_date)) AS Commuter_Trips FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) > 15 AND EXTRACT(HOUR FROM start_date) < 19 GROUP BY subscriber_type'

Morning commuter trips - start time between 7:00 and 8:59:
Waiting on bqjob_r509acd59cd109ea1_00000163bf2514ae_1 ... (0s) Current status: DONE   
+-----------------+----------------+
| subscriber_type | Commuter_Trips |
+-----------------+----------------+
| Subscriber      |         192117 |
| Customer        |           7878 |
+-----------------+----------------+
Afternoon commuter trips - start time between 16:00 and 18:59:
Waiting on bqjob_r579604f1136da4d2_00000163bf2521e0_1 ... (0s) Current status: DONE   
+-----------------+----------------+
| subscriber_type | Commuter_Trips |
+-----------------+----------------+
| Customer        |          32862 |
| Subscriber      |         266764 |
+-----------------+----------------+


In [38]:
print("Most popular starting points for morning and afternoon commuter trips:")
! bq query --use_legacy_sql=false --format=csv 'SELECT COUNT(start_station_id) AS Morning_Trips, start_station_name FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) < 9 AND EXTRACT(HOUR FROM start_date) > 6 AND duration_sec <= 1800 GROUP BY start_station_name ORDER BY COUNT(start_station_id) DESC LIMIT 5' > am_start.csv
! bq query --use_legacy_sql=false --format=csv 'SELECT COUNT(start_station_id) AS Afternoon_Trips, start_station_name FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) < 19 AND EXTRACT(HOUR FROM start_date) > 15 AND duration_sec <= 1800 GROUP BY start_station_name ORDER BY COUNT(start_station_id) DESC LIMIT 5' > pm_start.csv

Most popular starting points for morning and afternoon commuter trips:
Waiting on bqjob_r12b81a3d7df6b6f5_00000163bf0af132_1 ... (0s) Current status: DONE   

In [39]:
print("Most popular destinations for morning and afternoon commuter trips:")
! bq query --use_legacy_sql=false --format=csv 'SELECT COUNT(end_station_id) AS Morning_Trips, end_station_name FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) < 9 AND EXTRACT(HOUR FROM start_date) > 6 AND duration_sec <= 1800 GROUP BY end_station_name ORDER BY COUNT(end_station_id) DESC LIMIT 5' > am_end.csv
! bq query --use_legacy_sql=false --format=csv 'SELECT COUNT(end_station_id) AS Afternoon_Trips, end_station_name FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) < 19 AND EXTRACT(HOUR FROM start_date) > 15 AND duration_sec <= 1800 GROUP BY end_station_name ORDER BY COUNT(end_station_id) DESC LIMIT 5' > pm_end.csv

Most popular destinations for morning and afternoon commuter trips:
Waiting on bqjob_r1c7614b193f817ad_00000163bf0b79ab_1 ... (0s) Current status: DONE   

In [32]:
print("Most popular morning and afternoon commuter trips:")
! bq query --use_legacy_sql=false --format=csv 'SELECT CONCAT(start_station_name,"-",end_station_name) AS Trip, COUNT(CONCAT(start_station_name,"-",end_station_name)) AS Number_Trips FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) < 9 AND EXTRACT(HOUR FROM start_date) > 6 AND duration_sec <= 1800 GROUP BY trip ORDER BY COUNT(CONCAT(start_station_name,"-",end_station_name)) DESC LIMIT 5' > morning.csv
! bq query --use_legacy_sql=false --format=csv 'SELECT CONCAT(start_station_name,"-",end_station_name) AS Trip, COUNT(CONCAT(start_station_name,"-",end_station_name)) AS Number_Trips FROM `bigquery-public-data.san_francisco.bikeshare_trips` WHERE EXTRACT(HOUR FROM start_date) < 19 AND EXTRACT(HOUR FROM start_date) > 15 AND duration_sec <= 1800 GROUP BY trip ORDER BY COUNT(CONCAT(start_station_name,"-",end_station_name)) DESC LIMIT 5' > afternoon.csv

Most popular morning and afternoon commuter trips:
Waiting on bqjob_r7075e1c177612b5_00000163bf059811_1 ... (1s) Current status: DONE    

In [40]:
morning_trips=pd.read_csv('morning.csv')
afternoon_trips=pd.read_csv('afternoon.csv')
morning_start=pd.read_csv('am_start.csv')
afternoon_start=pd.read_csv('pm_start.csv')
morning_end=pd.read_csv('am_end.csv')
afternoon_end=pd.read_csv('pm_end.csv')

In [41]:
pd.set_option('display.max_colwidth', -1)

In [42]:
morning_trips

Trip  \
0  Harry Bridges Plaza (Ferry Building)-2nd at Townsend                                     
1  San Francisco Caltrain (Townsend at 4th)-Embarcadero at Folsom                           
2  San Francisco Caltrain (Townsend at 4th)-Harry Bridges Plaza (Ferry Building)            
3  San Francisco Caltrain (Townsend at 4th)-Temporary Transbay Terminal (Howard at Beale)   
4  San Francisco Caltrain (Townsend at 4th)-Steuart at Market                               

   Number_Trips  
0  3680          
1  2613          
2  2564          
3  2353          
4  2229

In [43]:
afternoon_trips

Trip  \
0  2nd at Townsend-Harry Bridges Plaza (Ferry Building)                                     
1  Embarcadero at Folsom-San Francisco Caltrain (Townsend at 4th)                           
2  Embarcadero at Sansome-Steuart at Market                                                 
3  Steuart at Market-San Francisco Caltrain (Townsend at 4th)                               
4  Temporary Transbay Terminal (Howard at Beale)-San Francisco Caltrain (Townsend at 4th)   

   Number_Trips  
0  4296          
1  4055          
2  3998          
3  3424          
4  3332

In [44]:
morning_start

Morning_Trips                             start_station_name
0  30303          San Francisco Caltrain (Townsend at 4th)     
1  21183          San Francisco Caltrain 2 (330 Townsend)      
2  15818          Harry Bridges Plaza (Ferry Building)         
3  14193          Temporary Transbay Terminal (Howard at Beale)
4  9762           Steuart at Market

In [45]:
afternoon_start

Afternoon_Trips                        start_station_name
0  14499            Townsend at 7th                         
1  14171            2nd at Townsend                         
2  14032            San Francisco Caltrain (Townsend at 4th)
3  13874            Embarcadero at Sansome                  
4  11386            Market at Sansome

In [46]:
morning_end

Morning_Trips                          end_station_name
0  16663          San Francisco Caltrain (Townsend at 4th)
1  11321          Townsend at 7th                         
2  11251          2nd at Townsend                         
3  8927           Market at Sansome                       
4  8763           Embarcadero at Sansome

In [47]:
afternoon_end

Afternoon_Trips                               end_station_name
0  45704            San Francisco Caltrain (Townsend at 4th)     
1  31626            San Francisco Caltrain 2 (330 Townsend)      
2  19016            Harry Bridges Plaza (Ferry Building)         
3  15482            Steuart at Market                            
4  15006            Temporary Transbay Terminal (Howard at Beale)

Per the assignment 4 instructions, I have refined my definitions of morning and afternoon trips to encompass 7:00 – 9:00 and 16:00 – 19:00, respectively.

The overall quantity of morning trips, represented by those with start times between 7:00 and 8:59, was 199,995. The overall quantity of afternoon trips, represented by those start times between 16:00 and 18:59, was 299,626.

I refined this further by checking the quantity of trips in each of these categories with durations of under 30 minutes. 197,646 of the 199,995 morning trips and 289,743 of the 299,626 afternoon trips fell into this category. This demonstrated that the overwhelming majority of trips within the hours selected could be considered commuter trips.

Next, I examined the breakdown of customer types for morning and afternoon trips. Subscribers accounted for 192,117 of the 199,995 morning trips and 266,764 of the 299,626 afternoon trips. It is clear that subscribers making commuter trips represented a large portion of the customer base in this dataset for trips made within commuting hours. The company running Bay Area Bikeshare would do well to reinforce and reward this behavior.

I also studied the top five starting and ending points for morning and afternoon commuter trips. This analysis gave more actionable results. The frequencies of the most common start and end points ranged from 8,763 to 45,704, representing a much more substantial portion of the overall trip totals. The top five starting and ending points for morning and afternoon commuter trips is displayed in the morning_start, morning_end, afternoon_start, and afternoon_end tables above. The morning starting points and afternoon ending points are most telling. The top five locations in these two sets match.

As a result of this analysis, my recommendations to the company running Bay Area Bikeshare would be to offer discounted rates for new subscribers who begin morning trips or end afternoon trips of up to 30 minutes in duration at any of the top five locations described above: San Francisco Caltrain (Townsend at 4th), San Francisco Caltrain 2 (330 Townsend), Harry Bridges Plaza (Ferry Building), Temporary Transbay Terminal (Howard at Beale), or Steuart at Market. The data do not give any indication of bike availability at these locations during peak times. I would recommend the company make certain to have sufficient stock at these locations during commuting hours, and plan to expand inventory to meet the new demand that will accompany the discounted rates.